tls/s2n_x509_validator.c

```diff
@@ -0,0 +1,450 @@
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "tls/s2n_config.h"
+#include "utils/s2n_asn1_time.h"
+#include "utils/s2n_safety.h"
+#include "tls/s2n_connection.h"
+
+#include "openssl/err.h"
+#include "openssl/asn1.h"
+
+#if !defined(OPENSSL_IS_BORINGSSL)
+#include "openssl/ocsp.h"
+#endif
+
+/* one day, boringssl, may add ocsp stapling support. Let's future proof this a bit by grabbing a definition
+ * that would have to be there when they add support */
+#if defined(OPENSSL_IS_BORINGSSL) && !defined(OCSP_RESPONSE_STATUS_SUCCESSFUL)
+#define S2N_OCSP_STAPLING_SUPPORTED 0
+#else
+#define S2N_OCSP_STAPLING_SUPPORTED 1
+#endif /* defined(OPENSSL_IS_BORINGSSL) && !defined(OCSP_RESPONSE_STATUS_SUCCESSFUL) */
+
+/* our friends at openssl love to make backwards incompatible changes */
+#if !defined(LIBRESSL_VERSION_NUMBER) && OPENSSL_VERSION_NUMBER >= 0x10100000L
+#define OCSP_GET_CERTS(a) OCSP_resp_get0_certs(a)
+#else
+#define OCSP_GET_CERTS(a) a->certs
+#endif
+
+uint8_t s2n_x509_ocsp_stapling_supported(void) {
+    return S2N_OCSP_STAPLING_SUPPORTED;
+}
+
+void s2n_x509_trust_store_init_empty(struct s2n_x509_trust_store *store) {
+    store->trust_store = NULL;
+}
+
+uint8_t s2n_x509_trust_store_has_certs(struct s2n_x509_trust_store *store) {
+    return store->trust_store ? (uint8_t) 1 : (uint8_t) 0;
+}
+
+int s2n_x509_trust_store_from_system_defaults(struct s2n_x509_trust_store *store) {
+    if(!store->trust_store) {
+        store->trust_store = X509_STORE_new();
+    }
+
+    int err_code = X509_STORE_set_default_paths(store->trust_store);
+
+    if (!err_code) {
+        s2n_x509_trust_store_wipe(store);
+        return -1;
+    }
+
+    X509_STORE_set_flags(store->trust_store, X509_VP_FLAG_DEFAULT);
+
+    return 0;
+}
+
+int s2n_x509_trust_store_from_ca_file(struct s2n_x509_trust_store *store, const char *ca_file, const char *path) {
+
+    if(!store->trust_store) {
+        store->trust_store = X509_STORE_new();
+    }
+
+    int err_code = X509_STORE_load_locations(store->trust_store, ca_file, path);
+
+    if (!err_code) {
+        s2n_x509_trust_store_wipe(store);
+        return -1;
+    }
+
+    /* It's a likely scenario if this function is called, a self-signed certificate is used, and that is was generated
+     * without a trust anchor. However if you call this function, the assumption is you trust ca_file or path and if a certificate
+     * is encountered that's in that path, it should be trusted. The following flag tells libcrypto to not care that the cert
+     * is missing a root anchor. */
+    unsigned long flags = X509_VP_FLAG_DEFAULT;
+
+#if defined(X509_V_FLAG_PARTIAL_CHAIN)
+    flags |=  X509_V_FLAG_PARTIAL_CHAIN;
+#else
+    flags |= 0x80000;
```

Comment:
Whats `0x80000`? Can we move it to a named constant?

Reply:
It's for ancient openssl not having the named constant X509_V_FLAG_PARTIAL_CHAIN, but still respecting the value.